Source/Core/InputCommon/InputConfig.cpp

```diff
@@ -117,3 +117,17 @@ bool InputConfig::ControllersNeedToBeCreated() const
 {
 	return m_controllers.empty();
 }
+
+bool InputConfig::IsControllerControlledBySomeTypeOfGamepadDevice(int index) const
+{
+	if (static_cast<size_t>(index) >= m_controllers.size())
+		return false;
+
+	const auto& controller = m_controllers.at(index).get()->default_device;
+
+	// Filter out anything which obviously not a gamepad
+	return !((controller.source == "Keyboard") // OSX Keyboard/Mouse
```

Comment:
What was particularly wrong with the more alignable and less likely to be overlooked (assuming I did the inversions correctly):

```cpp
return (controller.source != "Keyboard") &&                                     // OSX Keyboard/Mouse
       (controller.source != "XInput2")  &&                                     // Linux and BSD Keyboard/Mouse
       (controller.source != "Android" || controller.name != "Touchscreen") &&  // Android Touchscreen
       (controller.source != "DInput"  || controller.name != "Keyboard Mouse"); // Windows Keyboard/Mouse
```
?